File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/planning/promotion_controller.py
import inspect
import logging
from typing import Union, Dict, Any
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.db.models import QuerySet
from django.http import (HttpRequest, HttpResponse, HttpResponseRedirect, 
                         HttpResponsePermanentRedirect, JsonResponse, Http404)
from django.shortcuts import redirect, render, get_object_or_404
from .._helpers.error_handlers import (
    default_permission_denial,
    default_undefined_exception,
)
from .._helpers.http import get_redirect_url
from .._traits.controller import Controller
from ....Models.individuals.designation import Designation
from ....Models.individuals.employee import Employee
from ....Models.planning.promotion import Promotion
from ....Models.utils.utility import Utility

logger = logging.getLogger(__name__)


class PromotionController(Controller):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.logger = logger

    @classmethod
    def _check_permission(cls, request: HttpRequest, perm: str, json: bool = False) -> Union[Exception, bool]:
        self = cls()
        self.request = request
        try:
            self.authorize(perm)
            return True
        except Exception as e:
            return default_permission_denial(
                request,
                e,
                ref=f"{cls.__name__}::{inspect.currentframe().f_code.co_name}",
                logger=cls().logger,
                json={"error": "Permission denied."} if json else None,
            )

    def _get_user_promotions(self, request: HttpRequest) -> Union[HttpResponse, JsonResponse, None]:
        user = request.user
        qs = Promotion.objects.filter(created_by=user.creator_id())
        if user.type == "Employee":
            emp = Employee.objects.filter(user_id=user.id).first()
            if emp:
                qs = qs.filter(employee_id=emp.id)
        return qs

    def _pluck(self, qs: QuerySet, key: str, val: Any) -> Dict[Any, Any]:
        return dict(qs.values_list(key, val))

    def index(self, request: HttpRequest) -> Union[HttpResponse, JsonResponse, None]:
        REF=f"{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}"
        try:
            self._check_permission(request, "manage promotion")
            promotions = (
                self._get_user_promotions(request)
                .select_related("designation", "employee")
            )
            return render(request, "promotion/index.html", {"promotions": promotions})
        except PermissionDenied as e:
            return default_permission_denial(
                request,
                e,
                ref=REF,
                logger=self.logger,
                json={"error": "Permission denied."},
            )
        except Exception as e:
            return default_undefined_exception(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )

    def create(self, request: HttpRequest) -> Union[HttpResponse, JsonResponse, None]:
        REF=f"{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}"
        try:
            self._check_permission(request, "create promotion", json=True)
            user = request.user
            designations = self._pluck(
                Designation.objects.filter(created_by=user.creator_id()), "id", "name"
            )
            employees = self._pluck(
                Employee.objects.filter(created_by=user.creator_id()), "id", "name"
            )
            return render(
                request,
                "promotion/create.html",
                {"designations": designations, "employees": employees},
            )
        except PermissionDenied as e:
            return default_permission_denial(
                request,
                e,
                ref=REF,
                logger=self.logger,
                json={"error": "Permission denied."},
            )
        except Exception as e:
            return default_undefined_exception(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )

    def store(self, request: HttpRequest) -> Union[HttpResponse, JsonResponse, None]:
        REF=f"{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}"
        try:
            self._check_permission(request, "create promotion")
            data = request.POST
            required = ("employee_id", "designation_id", "promotion_title", "promotion_date")
            missing = [f for f in required if not data.get(f)]
            if missing:
                messages.error(
                    request,
                    f"{missing[0].replace('_',' ').capitalize()} is required",
                )
                return redirect(get_redirect_url(request))
            pmt_props = {}
            for k in (*required, 'description'):
              if k == 'description': setattr(pmt_props, k, data.get(k, ''))
              else: setattr(pmt_props, k, data.get(k))
            promotion = Promotion(
              **pmt_props,
              created_by=request.user.creator_id(),
            )
            promotion.save()
            settings = Utility.settings()
            if settings.get("promotion_sent") == 1:
                emp = Employee.objects.filter(id=promotion.employee_id).first()
                des = Designation.objects.filter(id=promotion.designation_id).first()
                promotion.designation = des.name if des else ""
                arr = {
                    "employee_name": emp.name if emp else "",
                    "promotion_designation": promotion.designation,
                    "promotion_title": promotion.promotion_title,
                    "promotion_date": promotion.promotion_date,
                }
                resp = Utility.send_email_template(
                    "promotion_sent", [emp.email] if emp else [], arr
                )
                msg = "Promotion successfully created."
                if resp and not resp.get("is_success") and resp.get("error"):
                    msg += f"<br><span class='text-danger'>{resp['error']}</span>"
                messages.success(request, msg)
            else:
                messages.success(request, "Promotion successfully created.")
            return redirect("promotion_index")
        except PermissionDenied as e:
            return default_permission_denial(
                request,
                e,
                ref=REF,
                logger=self.logger,
                json={"error": "Permission denied."},
            )
        except Exception as e:
            return default_undefined_exception(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )

    def show(self, request: HttpRequest, promotion_id:str) -> Union[Http404, HttpResponseRedirect, HttpResponsePermanentRedirect]:
        try:
          get_object_or_404(Promotion, pk=promotion_id)
          return redirect(get_redirect_url(request) if request.META.get('HTTP_REFERER') else "promotion_index")
        except Exception as e:
            return default_undefined_exception(
                request,
                e,
                ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
                logger=self.logger,
            )

    def edit(self, request: HttpRequest, promotion_id:str) -> Union[HttpResponse, JsonResponse, None]:
        REF = f"{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}"
        try:
            self._check_permission(request, "edit promotion", json=True)
            promotion = Promotion.objects.filter(
                id=promotion_id, created_by=request.user.creator_id()
            ).first()
            if not promotion:
                raise PermissionDenied
            user = request.user
            designations = self._pluck(
                Designation.objects.filter(created_by=user.creator_id()), "id", "name"
            )
            employees = self._pluck(
                Employee.objects.filter(created_by=user.creator_id()), "id", "name"
            )
            return render(
                request,
                "promotion/edit.html",
                {
                    "promotion": promotion,
                    "designations": designations,
                    "employees": employees,
                },
            )
        except PermissionDenied as e:
            return default_permission_denial(
                request,
                e,
                ref=REF,
                logger=self.logger,
                json={"error": "Permission denied."},
            )
        except Exception as e:
            return default_undefined_exception(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )

    def update(self, request: HttpRequest, promotion_id:str) -> Union[HttpResponse, JsonResponse, None]:
        REF = f"{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}"
        try:
            self._check_permission(request, "edit promotion")
            promotion = Promotion.objects.filter(
                id=promotion_id, created_by=request.user.creator_id()
            ).first()
            if not promotion:
                messages.error(request, "Permission denied.")
                return redirect(get_redirect_url(request))
            data = request.POST
            required = ("employee_id", "designation_id", "promotion_title", "promotion_date")
            missing = [f for f in required if not data.get(f)]
            if missing:
                messages.error(
                    request,
                    f"{missing[0].replace('_',' ').capitalize()} is required",
                )
                return redirect(get_redirect_url(request))
            for field in (
                "employee_id",
                "designation_id",
                "promotion_title",
                "promotion_date",
                "description",
            ):
                if data.get(field) is not None:
                    setattr(promotion, field, data.get(field))
            promotion.save()
            messages.success(request, "Promotion successfully updated.")
            return redirect("promotion_index")
        except PermissionDenied as e:
            return default_permission_denial(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )
        except Exception as e:
            return default_undefined_exception(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )

    def destroy(self, request: HttpRequest, promotion_id:str) -> Union[HttpResponse, JsonResponse, None]:
        REF = f"{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}"
        try:
            self._check_permission(request, "delete promotion")
            promotion = Promotion.objects.filter(
                id=promotion_id, created_by=request.user.creator_id()
            ).first()
            if not promotion:
                raise PermissionDenied
            promotion.delete()
            messages.success(request, "Promotion successfully deleted.")
            return redirect("promotion_index")
        except PermissionDenied as e:
            return default_permission_denial(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )
        except Exception as e:
            return default_undefined_exception(
                request,
                e,
                ref=REF,
                logger=self.logger,
            )

File: _inc/laravel/app/Http/Controllers/Planning/PromotionController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    DatabaseConstants,
    PermissionsConstants,
    UsersConstants,
    ViewsConstants
};
use App\Models\{Designation, Employee, Promotion, Utility};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Contracts\View\View;
use Illuminate\Http\{JsonResponse, RedirectResponse, Request};
use Illuminate\Support\Arr;
use Illuminate\Support\Facades\{
    Auth,
    Log,
    Mail,
    Route,
    Validator,
    View as ViewFacade
};
use Symfony\Component\HttpFoundation\Response;

class PromotionController extends Controller
{
    use ChecksLogin;
    use ChecksPermissions;

    private const REDIRECT_INDEX = ViewsConstants::PRM . '.index';

    public function index(Request $request): View|RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $req    = $request;

        return $this->measureProfile($action, function () use ($req, $action, $method, $class) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($resp = self::guard($req, PermissionsConstants::MNG_PRM, self::REDIRECT_INDEX)) !== true) return $resp;

            Log::info("[{$class}::{$action}] start");
            try {
                $buildStart = microtime(true);
                $promotions = Promotion::query()
                    ->where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->when(
                        strtolower($user[UsersConstants::COL_TP]) === 'employee',
                        function ($q) use ($user) {
                            $empId = Employee::where(UsersConstants::COL_USER_ID, $user?->id)->value('id');
                            return $q->where(UsersConstants::COL_EMP_ID, $empId);
                        }
                    )
                    ->with(['designation', 'employee']);
                $this->logExecutionTime($buildStart, $action, 'buildPromotionQuery');

                $fetchStart = microtime(true);
                $promotions = $promotions->get();
                $this->logExecutionTime($fetchStart, $action, 'fetchPromotions');

                Log::info("[{$class}::{$action}] ready", ['count' => $promotions->count()]);
                return $this->renderViewChecked(self::REDIRECT_INDEX, compact('promotions'), $action);
            } catch (\Throwable $e) {
                Log::debug("[{$class}::{$action}] debug error context", [
                    'route'   => Route::getCurrentRoute()?->getName(),
                    'method'  => $method,
                    'class'   => $class,
                    'user_id' => $user?->id ?? null,
                    'message' => $e->getMessage(),
                    'file'    => $e->getFile(),
                    'line'    => $e->getLine(),
                ]);
                Log::error("[{$class}::{$action}] error", ['error' => $e->getMessage()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, route(self::REDIRECT_INDEX));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
    }

    public function create(Request $request): View|RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $req    = $request;
        $viewPath = ViewsConstants::PRM . '.' . $action;

        return $this->measureProfile($action, function () use ($req, $action, $method, $class, $viewPath) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($resp = self::guard($req, 'create promotion', self::REDIRECT_INDEX)) !== true) return $resp;

            Log::info("[{$class}::{$action}] start");
            try {
                $desStart = microtime(true);
                $designations = Designation::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck('name', 'id');
                $this->logExecutionTime($desStart, $action, 'pluckDesignations');

                $empStart = microtime(true);
                $employees = Employee::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck('name', 'id');
                $this->logExecutionTime($empStart, $action, 'pluckEmployees');

                Log::info("[{$class}::{$action}] ready", [
                    'designations' => count($designations),
                    'employees'    => count($employees),
                ]);

                return $this->renderViewChecked($viewPath, [
                    'designations' => $designations,
                    'employees'    => $employees,
                ], $action);
            } catch (\Throwable $e) {
                Log::debug("[{$class}::{$action}] debug error context", [
                    'route'   => Route::getCurrentRoute()?->getName(),
                    'method'  => $method,
                    'class'   => $class,
                    'user_id' => $user?->id ?? null,
                    'message' => $e->getMessage(),
                    'file'    => $e->getFile(),
                    'line'    => $e->getLine(),
                ]);
                Log::error("[{$class}::{$action}] error", ['error' => $e->getMessage()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, route(self::REDIRECT_INDEX));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
    }

    public function store(Request $request): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $req    = $request;

        return $this->measureProfile($action, function () use ($req, $action, $method, $class) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($resp = self::guard($req, 'create promotion', self::REDIRECT_INDEX)) !== true) return $resp;

            Log::info("[{$class}::{$action}] start");
            try {
                $valStart = microtime(true);
                $validator = Validator::make($req->all(), [
                    UsersConstants::COL_EMP_ID => 'required',
                    'designation_id'           => 'required',
                    'promotion_title'          => 'required',
                    'promotion_date'           => 'required',
                ]);
                $this->logExecutionTime($valStart, $action, 'buildValidator');

                $failCheckStart = microtime(true);
                if ($validator->fails()) {
                    $this->logExecutionTime($failCheckStart, $action, 'validatorFailsCheck');
                    Log::debug("[{$class}::{$action}] validation failed", [
                        'first_error' => $validator->errors()->first(),
                        'route'       => Route::getCurrentRoute()?->getName(),
                        'method'      => $method,
                        'class'       => $class,
                    ]);
                    return redirect()->back()->with('error', $validator->errors()->first());
                }

                $data = Arr::only($req->all(), [
                    UsersConstants::COL_EMP_ID,
                    'designation_id',
                    'promotion_title',
                    'promotion_date',
                ]);
                $data['description'] = $req->description ?? '';
                $data[DatabaseConstants::COL_TABLE_CREATOR] = $user?->creatorId();

                $createStart = microtime(true);
                $promotion = Promotion::create($data);
                $this->logExecutionTime($createStart, $action, 'createPromotion');

                $settingsStart = microtime(true);
                $settings = Utility::settings();
                $this->logExecutionTime($settingsStart, $action, 'loadSettings');

                if (($settings['promotion_sent'] ?? 0) == 1) {
                    $empFetchStart = microtime(true);
                    $employee    = Employee::find($promotion->employee_id);
                    $designation = Designation::find($promotion->designation_id);
                    $this->logExecutionTime($empFetchStart, $action, 'fetchEmployeeDesignation');

                    $promotionArr = [
                        'employee_name'         => $employee->name,
                        'promotion_designation' => $designation->name,
                        'promotion_title'       => $promotion->promotion_title,
                        'promotion_date'        => $promotion->promotion_date,
                    ];

                    $mailStart = microtime(true);
                    $resp = Utility::sendEmailTemplate('promotion_sent', [$employee->email], $promotionArr);
                    $this->logExecutionTime($mailStart, $action, 'sendPromotionEmail');

                    $msg = __('Promotion successfully created.');
                    $msg .= (!empty($resp) && $resp['is_success'] === false && !empty($resp['error']))
                        ? "<br><span class=\"text-danger\">{$resp['error']}</span>"
                        : '';

                    return redirect()->route(self::REDIRECT_INDEX)->with('success', $msg);
                }

                return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Promotion successfully created.'));
            } catch (\Throwable $e) {
                Log::debug("[{$class}::{$action}] debug error context", [
                    'route'      => Route::getCurrentRoute()?->getName(),
                    'method'     => $method,
                    'class'      => $class,
                    'input_keys' => array_keys($req->all()),
                    'user_id'    => $user?->id ?? null,
                    'message'    => $e->getMessage(),
                    'file'       => $e->getFile(),
                    'line'       => $e->getLine(),
                ]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, route(self::REDIRECT_INDEX));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
    }

    public function show(Request $request, Promotion $promotion): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $req    = $request;

        return $this->measureProfile($action, function () use ($req, $promotion, $action, $method, $class) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            Log::info("[{$class}::{$action}] start", ['promotion_id' => $promotion->id]);
            try {
                $redirStart = microtime(true);
                $resp = redirect()->route(self::REDIRECT_INDEX);
                $this->logExecutionTime($redirStart, $action, 'buildRedirect');

                Log::info("[{$class}::{$action}] redirecting to index");
                return $resp;
            } catch (\Throwable $e) {
                Log::debug("[{$class}::{$action}] debug error context", [
                    'promotion_id' => $promotion->id,
                    'route'        => Route::getCurrentRoute()?->getName(),
                    'method'       => $method,
                    'class'        => $class,
                    'message'      => $e->getMessage(),
                    'file'         => $e->getFile(),
                    'line'         => $e->getLine(),
                ]);
                Log::error("[{$class}::{$action}] error", ['error' => $e->getMessage()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, route(self::REDIRECT_INDEX));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'promotion_id' => $promotion->id]);
    }

    public function edit(Request $request, Promotion $promotion): View|RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $req    = $request;
        $viewPath = ViewsConstants::PRM . '.edit';

        return $this->measureProfile($action, function () use ($req, $promotion, $action, $method, $class, $viewPath) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($resp = self::guard($req, 'edit promotion', self::REDIRECT_INDEX)) !== true) return $resp;
            if ($promotion->created_by !== $user?->creatorId())
                return response()->json(['error' => __('Permission denied.')], Response::HTTP_UNAUTHORIZED);

            Log::info("[{$class}::{$action}] start", ['promotion_id' => $promotion->id]);
            try {
                $desStart = microtime(true);
                $designations = Designation::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck('name', 'id');
                $this->logExecutionTime($desStart, $action, 'pluckDesignations');

                $empStart = microtime(true);
                $employees = Employee::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck('name', 'id');
                $this->logExecutionTime($empStart, $action, 'pluckEmployees');

                Log::info("[{$class}::{$action}] ready", [
                    'promotion_id'  => $promotion->id,
                    'designations'  => count($designations),
                    'employees'     => count($employees),
                ]);

                return $this->renderViewChecked($viewPath, [
                    'promotion'    => $promotion,
                    'designations' => $designations,
                    'employees'    => $employees,
                ], $action);
            } catch (\Throwable $e) {
                Log::debug("[{$class}::{$action}] debug error context", [
                    'route'        => Route::getCurrentRoute()?->getName(),
                    'method'       => $method,
                    'class'        => $class,
                    'promotion_id' => $promotion->id,
                    'message'      => $e->getMessage(),
                    'file'         => $e->getFile(),
                    'line'         => $e->getLine(),
                ]);
                Log::error("[{$class}::{$action}] error", ['error' => $e->getMessage()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, route(self::REDIRECT_INDEX));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'promotion_id' => $promotion->id]);
    }

    public function update(Request $request, Promotion $promotion): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $req    = $request;

        return $this->measureProfile($action, function () use ($req, $promotion, $action, $method, $class) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($resp = self::guard($req, 'edit promotion', self::REDIRECT_INDEX)) !== true) return $resp;
            if ($promotion->created_by !== $user?->creatorId())
                return defaultPermissionDenial($req, new AuthorizationException($promotion->getKey()), $class . '::' . $action, route(self::REDIRECT_INDEX));

            Log::info("[{$class}::{$action}] start", ['promotion_id' => $promotion->id]);
            try {
                $valStart = microtime(true);
                $validator = Validator::make($req->all(), [
                    UsersConstants::COL_EMP_ID => 'required',
                    'designation_id'           => 'required',
                    'promotion_title'          => 'required',
                    'promotion_date'           => 'required',
                ]);
                $this->logExecutionTime($valStart, $action, 'buildValidator');

                $failCheckStart = microtime(true);
                if ($validator->fails()) {
                    $this->logExecutionTime($failCheckStart, $action, 'validatorFailsCheck');
                    Log::debug("[{$class}::{$action}] validation failed", [
                        'first_error'  => $validator->errors()->first(),
                        'route'        => Route::getCurrentRoute()?->getName(),
                        'method'       => $method,
                        'class'        => $class,
                        'promotion_id' => $promotion->id,
                    ]);
                    return redirect()->back()->with('error', $validator->errors()->first());
                }

                $data = Arr::only($req->all(), [
                    UsersConstants::COL_EMP_ID,
                    'designation_id',
                    'promotion_title',
                    'promotion_date',
                ]);
                $data['description'] = $req->description ?? '';

                $updStart = microtime(true);
                $promotion->update($data);
                $this->logExecutionTime($updStart, $action, 'updatePromotion');

                Log::info("[{$class}::{$action}] updated", ['promotion_id' => $promotion->id]);
                return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Promotion successfully updated.'));
            } catch (\Throwable $e) {
                Log::debug("[{$class}::{$action}] debug error context", [
                    'promotion_id' => $promotion->id,
                    'route'        => Route::getCurrentRoute()?->getName(),
                    'method'       => $method,
                    'class'        => $class,
                    'input_keys'   => array_keys($req->all()),
                    'user_id'      => $user?->id ?? null,
                    'message'      => $e->getMessage(),
                    'file'         => $e->getFile(),
                    'line'         => $e->getLine(),
                ]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, route(self::REDIRECT_INDEX));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'promotion_id' => $promotion->id]);
    }

    public function destroy(Request $request, Promotion $promotion): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $req    = $request;

        return $this->measureProfile($action, function () use ($req, $promotion, $action, $method, $class) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($resp = self::guard($req, 'delete promotion', self::REDIRECT_INDEX)) !== true) return $resp;
            if ($promotion->created_by !== $user?->creatorId())
                return defaultPermissionDenial($req, new AuthorizationException($promotion->getKey()), $class . '::' . $action, route(self::REDIRECT_INDEX));

            Log::info("[{$class}::{$action}] start", ['promotion_id' => $promotion->id]);
            try {
                $delStart = microtime(true);
                $promotion->delete();
                $this->logExecutionTime($delStart, $action, 'deletePromotion');

                Log::info("[{$class}::{$action}] deleted", ['promotion_id' => $promotion->id]);
                return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Promotion successfully deleted.'));
            } catch (\Throwable $e) {
                Log::debug("[{$class}::{$action}] debug error context", [
                    'promotion_id' => $promotion->id,
                    'route'        => Route::getCurrentRoute()?->getName(),
                    'method'       => $method,
                    'class'        => $class,
                    'user_id'      => $user?->id ?? null,
                    'message'      => $e->getMessage(),
                    'file'         => $e->getFile(),
                    'line'         => $e->getLine(),
                ]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, route(self::REDIRECT_INDEX));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'promotion_id' => $promotion->id]);
    }

    /**
     * Centralized view existence + render with timings.
     */
    private function renderViewChecked(string $viewPath, array $data, string $action): View|RedirectResponse
    {
        $existsStart = microtime(true);
        $exists = ViewFacade::exists($viewPath);
        $this->logExecutionTime($existsStart, $action, 'viewExistsCheck');

        if (!$exists) {
            Log::warning("[renderViewChecked] view missing", ['view' => $viewPath]);
            return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        }

        $renderStart = microtime(true);
        $response = view($viewPath, $data);
        $this->logExecutionTime($renderStart, $action, 'viewRender');

        return $response;
    }
}
